packages/playwright-core/src/server/injected/consoleApi.ts

```diff
@@ -14,40 +14,54 @@
-  class Locator {
-    selector: string;
-    element: Element | undefined;
-    elements: Element[];
+const selectorSymbol = Symbol('selector');
+const injectedScriptSymbol = Symbol('injectedScript');
 
-    constructor(selector: string, options?: { hasText?: string | RegExp, has?: Locator }) {
-      this.selector = selector;
-      if (options?.hasText)
-        this.selector += ` >> internal:has-text=${escapeForTextSelector(options.hasText, false)}`;
-      if (options?.has)
-        this.selector += ` >> internal:has=` + JSON.stringify(options.has.selector);
-      const parsed = injectedScript.parseSelector(this.selector);
+class Locator {
+  element: Element | undefined;
+  elements: Element[] | undefined;
+
+  constructor(injectedScript: InjectedScript, selector: string, options?: { hasText?: string | RegExp, has?: Locator }) {
+    (this as any)[selectorSymbol] = selector;
+    (this as any)[injectedScriptSymbol] = injectedScript;
+    if (options?.hasText)
+      selector += ` >> internal:has-text=${escapeForTextSelector(options.hasText, false)}`;
+    if (options?.has)
+      selector += ` >> internal:has=` + JSON.stringify((options.has as any)[selectorSymbol]);
+    if (selector) {
+      const parsed = injectedScript.parseSelector(selector);
       this.element = injectedScript.querySelector(parsed, document, false);
       this.elements = injectedScript.querySelectorAll(parsed, document);
     }
-
-    locator(selector: string, options?: { hasText: string | RegExp, has?: Locator }): Locator {
-      return new Locator(this.selector ? this.selector + ' >> ' + selector : selector, options);
-    }
+    const selectorBase = selector;
+    const self = this as any;
+    self.locator = (selector: string, options?: { hasText?: string | RegExp, has?: Locator }): Locator => {
+      return new Locator(injectedScript, selectorBase ? selectorBase + ' >> ' + selector : selector, options);
+    };
+    self.getByTestId = (testId: string): Locator => self.locator(getByTestIdSelector('data-testid', testId));
+    self.getByAltText = (text: string | RegExp, options?: { exact?: boolean }): Locator => self.locator(getByAltTextSelector(text, options));
+    self.getByLabel = (text: string | RegExp, options?: { exact?: boolean }): Locator => self.locator(getByLabelSelector(text, options));
+    self.getByPlaceholder = (text: string | RegExp, options?: { exact?: boolean }): Locator => self.locator(getByPlaceholderSelector(text, options));
+    self.getByText = (text: string | RegExp, options?: { exact?: boolean }): Locator => self.locator(getByTextSelector(text, options));
+    self.getByTitle = (text: string | RegExp, options?: { exact?: boolean }): Locator => self.locator(getByTitleSelector(text, options));
+    self.getByRole = (role: string, options: ByRoleOptions = {}): Locator => self.locator(getByRoleSelector(role, options));
   }
-  return new Locator(initial, options);
 }
 
 type ConsoleAPIInterface = {
```

Comment:
This type is now missing `getBy...` and `locator` methods. Do we even need it?

Reply:
I don't think we need it, but it does not seem to hurt either.